wsnet/dial.go

```diff
@@ -136,6 +140,10 @@ func (d *Dialer) negotiate() (err error) {
 				_ = connCloser.Close()
 			}
 			d.connClosers = make([]io.Closer, 0)
+			if !d.closed {
+				d.closed = true
+				close(d.closedChan)
+			}
```

Comment:
Instead of having a bool have a select with a default:
```go
select {
  case <-d.closedChan:
    return
  default:
}

close(d.closedChan)
```